### PR TITLE
Convert zero-inflated distributions to Mixture subclasses

### PR DESCRIPTION
**Pull request by fonnesbeck**
Added the ZIB distribution to `discrete.py`, along with tests.

### CODE REVIEW

File: pymc3/tests/test_distributions.py [hunk 1]
```diff
@@ -591,6 +591,10 @@ def test_zeroinflatednegativebinomial(self):
         self.checkd(ZeroInflatedNegativeBinomial, Nat,
                     {'mu': Rplusbig, 'alpha': Rplusbig, 'psi': Unit})
 
+    def test_zeroinflatedbinomial(self):
+        self.checkd(ZeroInflatedBinomial, Nat,
```
Comment (junpenglao) [inline pymc3/tests/test_distributions.py]: You forgot to import `ZeroInflatedBinomial` in the test_distributions.py

**Comment by twiecki** [comment]
Too bad we still haven't figured out how to do this with the Mixture class.

**Comment by fonnesbeck** [comment]
Let me look into it. Might not be worth the overhead just to use `Mixture`, but I will see if that's true.

**Comment by fonnesbeck** [comment]
Actually, its pretty easy. Hold off on merging this for the moment; I will convert all the zero-inflated distributions over and move them to `distributions/mixture`.

**Comment by twiecki** [comment]
Is it? I think we tried it but it wasn't easy to define a 0-distribution. CC @AustinRochford

**Comment by fonnesbeck** [comment]
Just used a `Constant.dist(0)`. 

[Here is the zero-inflated Poisson in action](https://gist.github.com/fonnesbeck/f33ae5865a262f049ff5255b6a5b2996).

**Comment by twiecki** [comment]
https://github.com/pymc-devs/pymc3/pull/1459

**Comment by fonnesbeck** [comment]
And [here is zero-inflated binomial](https://gist.github.com/fonnesbeck/86cd95c35eee033283c8caf47530e4f6). Seems to work. Let me know if I am missing something (I very well could be).

**Comment by fonnesbeck** [comment]
Please test.

File: pymc3/distributions/mixture.py [hunk 2]
```diff
@@ -139,6 +142,150 @@ def random_choice(*args, **kwargs):
...
+        Expected proportion of Poisson variates (0 < psi < 1)
+    theta : float
+        Expected number of occurrences during the given interval
+        (theta >= 0).
+
+
+    """
+
+    def __init__(self, psi, theta, *args, **kwargs):
+        self.theta = theta = tt.as_tensor_variable(theta)
```
Comment (junpenglao) [inline pymc3/distributions/mixture.py]: you should also cast `psi` into tensor (same below), although might not necessary after #2233 is merge

Comment (fonnesbeck) [reply pymc3/distributions/mixture.py]: That happens in the superclass init (Line 48)

**Comment by junpenglao** [comment]
Found a case that breaks it but not the old one.
https://gist.github.com/junpenglao/f71609fbd3b835d1bdc67e323fb7c141

**Comment by junpenglao** [comment]
The old one is far from perfect as well: because of the `tt.switch`, we can not use ADVI init (returns NaN all the time)

**Comment by fonnesbeck** [comment]
@junpenglao I would not have expected that to work. As I have implemented it, this is a two-component mixture; zeros and Binomial/Poisson/NB. That's the standard implementation of a zero-inflated distribution. We can make this explicit in the docstring, and one should be able to code up a custom Mixture for more obscure scenarios.

**Comment by fonnesbeck** [comment]
Docstrings are now explicit about 2-component mixture.

**Comment by fonnesbeck** [comment]
@junpenglao (also, you have misspecified the model slightly. You have 3 components (zeros and two poissons) but have only passed a Dirichlet of size 2)

**Comment by junpenglao** [comment]
Actually I was intended to model two different sessions, each session is a `ZeroInflatedPossion` with different mixture `psi`. In the notebook it is the same as if I model them separately but you can see they can be extended to a hierarchical model.

**Comment by fonnesbeck** [comment]
Yeah, that's a more complicated model than this was intended to represent.

**Comment by twiecki** [comment]
That looks great! Assigning @AustinRochford if he has time.

**Comment by fonnesbeck** [comment]
Since these are always 2-component mixtures, I've changed the `psi` to import a scalar rather than a vector (e.g. a Beta rather than a Dirichlet). This is consistent with the API of the existing zero-inflateds.

**Comment by fonnesbeck** [comment]
Seems to have a problem generating random draws with a multidimensional size. Not sure why.

**Comment by AustinRochford** [comment]
Looks good overall.  I had similar problems matching `scipy` and with random draws when I tried this before.  Will give some thought as to possible problems.

**Comment by AustinRochford** [comment]
Passing `broadcast_conditions=False` to `bound` in `Mixture`'s `logp` method fixes the broken ZIP and ZINB tests for me, but the ZIB test is still failing.

This was the reason for `broadcast_conditions`, so I'll add that in a separate PR and keep investigating the ZIB failure.

File: pymc3/distributions/mixture.py [hunk 3]
```diff
@@ -139,6 +142,156 @@ def random_choice(*args, **kwargs):
...
+    ========  ==========================
+    Support   :math:`x \in \mathbb{N}_0`
+    Mean      :math:`(1 - \psi) n p`
+    Variance  :math:`(1-\psi) n p [1 - p(1 - \psi n)].`
+    ========  ==========================
+
+    Parameters
+    ----------
+    psi : float
+        Expected proportion of Poisson variates (0 < psi < 1)
```
Comment (AustinRochford) [inline pymc3/distributions/mixture.py:217]: Should say Binomial

**Comment by AustinRochford** [comment]
@fonnesbeck the `broadcast_conditions` change is merged to master, rebasing should get more (but not all) tests to pass.

**Comment by fonnesbeck** [comment]
Well, I rebased instead of merged, so this branch is irrevocably messed up. Closing in favor of #2246

**Pull request closed without merging**